anesthetic/weighted_pandas.py

```diff
@@ -14,28 +15,59 @@ class _WeightedObject(object):
     def __init__(self, *args, **kwargs):
         weights = kwargs.pop('weights', None)
         super().__init__(*args, **kwargs)
-        self.weights = weights
+        if weights is not None:
+            self.set_weights(weights, inplace=True)
 
-    @property
-    def weights(self):
-        """Sample weights."""
-        if self.index.nlevels == 1:
-            return np.ones_like(self.index)
-        else:
-            return self.index.get_level_values('weights').to_numpy()
+    def isweighted(self, axis=0):
+        """Determine if weights are actually present."""
+        return 'weights' in self._get_axis(axis).names
 
-    @weights.setter
-    def weights(self, weights):
-        if weights is not None:
-            self.index = [self.index.get_level_values(0), weights]
-            self.index.set_names(['#', 'weights'], inplace=True)
+    def get_weights(self, axis=0):
+        """Retrieve sample weights from an axis."""
+        if self.isweighted(axis):
+            return self._get_axis(axis).get_level_values('weights').to_numpy()
+        else:
+            return np.ones_like(self._get_axis(axis))
+
+    def set_weights(self, weights, axis=0, inplace=False):
+        """Set sample weights along an axis."""
+        if weights is None:
+            if self.isweighted(axis=axis):
+                result = self.droplevel('weights', axis=axis)
+            elif inplace:
+                result = self
+            else:
+                result = self.copy()
+        else:
+            result = self.set_axis([self._get_axis(axis).get_level_values(name)
+                                    for name in self._get_axis(axis).names
+                                    if name != 'weights'] + [weights],
+                                   axis=axis)
+            names = result._get_axis(axis).names[:-1] + ['weights']
```

Comment:
As far as I can tell, nothing assumes that weight is the last index. In the above function `set_weights` puts it there, but nothing else assumes it (e.g. calling `droplevel('weights')` or `get_weights`. For indexing it is much more convenient to have weights as the last. However, it might be surprising to users if `set_weights` reordered a multiindex which they had been manipulating, so I have adjusted the function so

a) If weights is present, it keeps the same location in the multiindex
b) If weights is not present, it is appended last by default (for convenience)
c) the result, self syntax is more consistent with usual inplace optional functions.

Reply:
Tests now added to check this behaviour, as well as a `level` kwarg if you want to insert weights at a specific level (defaulting to the back).